lib/cdo/aws/dms.rb

```diff
@@ -65,5 +67,114 @@ def self.tasks(file)
+    def self.replication_task_status(replication_task_arn)
+      task = OpenStruct.new(arn: replication_task_arn)
+
+      dms_client = Aws::DatabaseMigrationService::Client.new
+      replication_task = dms_client.describe_replication_tasks(
+        {
+          filters: [
+            {
+              name: 'replication-task-arn',
+              values: [replication_task_arn]
+            }
+          ]
+        },
+        max_records: 1,
+        without_settings: true
+      ).replication_tasks[0]
+
+      # Collect the replication task attributes that identify overall replication status.
+      task.status = replication_task.status
+      task.last_failure_message = replication_task.last_failure_message
+      task.stop_reason = replication_task.stop_reason
+      task.start_date = replication_task.replication_task_start_date
+      task.full_load_progress_percent = replication_task.replication_task_stats.full_load_progress_percent
+      task.tables_loaded = replication_task.replication_task_stats.tables_loaded
+      task.tables_loading = replication_task.replication_task_stats.tables_loading
+      task.tables_queued = replication_task.replication_task_stats.tables_queued
+      task.tables_errored = replication_task.replication_task_stats.tables_errored
+
+      task.table_statistics = dms_client.describe_table_statistics(
+        {
+          replication_task_arn: replication_task_arn
+        }
+      ).table_statistics
+
+      return task
+    end
+
+    # Start a replication task and wait until it completes, raising an error if the task did not complete within a
+    # configurable time period or did not complete successfully.
+    # @param replication_task_arn [String]
+    # @param max_attempts [Integer] Number of times to check whether task has completed successfully before failing.
+    # @param delay [Integer] Number of seconds to wait between checking task status.
+    def self.start_replication_task(replication_task_arn, max_attempts, delay)
+      CDO.log.info "Starting DMS Replication Task: #{replication_task_arn}"
+      task = replication_task_status(replication_task_arn)
+      dms_client = Aws::DatabaseMigrationService::Client.new
+      dms_client.start_replication_task(
+        {
+          replication_task_arn: replication_task_arn,
+          # If a Replication Task has never been executed before, start it, otherwise, reload it.
+          start_replication_task_type: task.status == 'ready' ? 'start-replication' : 'reload-target'
+        }
+      )
+
+      wait_until_replication_task_completed(replication_task_arn, max_attempts, delay)
+
+      CDO.log.info "DMS Task Completed Successfully: #{replication_task_arn}"
+    rescue StandardError => error
+      CDO.log.info "Error executing DMS Replication Task #{replication_task_arn} - #{error.message}"
+      raise error
+    end
+
+    def self.wait_until_replication_task_completed(replication_task_arn, max_attempts, delay)
```

Comment:
I know you mentioned this before, but can you remind me why it's not possible to use the existing waiters here (e.g., `ReplicationTaskRunning` followed by `ReplicationTaskStopped`)?

Reply:
Will and I took a look at this offline.  These waiters exist to assist in detecting whether a particular DMS API operation has completed successfully.  The  ReplicationTaskStopped waiter can only be used to detect that an explicitly issued stop task operation has completed, not that a task that was started has finished.